src/load_cell_endstop.c

```diff
@@ -0,0 +1,297 @@
+                , int32_t tare_counts, uint32_t trigger_grams
+                , fixedQ1_t grams_per_count)
+{
+    if (!(safety_counts_max >= safety_counts_min)) {
+        shutdown("Safety range reversed");
+    }
+    if (trigger_grams > MAX_TRIGGER_GRAMS) {
+        shutdown("trigger_grams too large");
+    }
+    // grams_per_count must be a positive fraction in Q1 format
+    const fixedQ1_t one = 1 << FIXEDQ1_FRAC_BITS;
+    if (grams_per_count < 0 || grams_per_count >= one) {
+        shutdown("grams_per_count is invalid");
+    }
+    lce->safety_counts_min = safety_counts_min;
+    lce->safety_counts_max = safety_counts_max;
+    lce->tare_counts = tare_counts;
+    lce->trigger_grams = trigger_grams;
+    lce->trigger_grams_fixed = trigger_grams << FIXEDQ16_FRAC_BITS;
+    lce->grams_per_count = grams_per_count;
+    reset_filter_state(lce->sf);
+}
+
+// Create a load_cell_endstop
+void
+command_config_load_cell_endstop(uint32_t *args)
+{
+    struct load_cell_endstop *lce = oid_alloc(args[0]
+                            , command_config_load_cell_endstop, sizeof(*lce));
+    lce->flags = 0;
+    lce->trigger_ticks = 0;
+    lce->watchdog_max = 0;
+    lce->watchdog_count = 0;
+    lce->sf = sos_filter_oid_lookup(args[1]);
+    set_endstop_range(lce, 0, 0, 0, 0, 0);
+}
+DECL_COMMAND(command_config_load_cell_endstop, "config_load_cell_endstop"
+                                               " oid=%c sos_filter_oid=%c");
+
+// Lookup a load_cell_endstop
+struct load_cell_endstop *
+load_cell_endstop_oid_lookup(uint8_t oid)
+{
+    return oid_lookup(oid, command_config_load_cell_endstop);
+}
+
+// Set the triggering range and tare value
+void
+command_set_range_load_cell_endstop(uint32_t *args)
+{
+    struct load_cell_endstop *lce = load_cell_endstop_oid_lookup(args[0]);
+    set_endstop_range(lce, args[1], args[2], args[3], args[4]
+                , (fixedQ16_t)args[5]);
+}
+DECL_COMMAND(command_set_range_load_cell_endstop, "set_range_load_cell_endstop"
+    " oid=%c safety_counts_min=%i safety_counts_max=%i tare_counts=%i"
+    " trigger_grams=%u grams_per_count=%i");
+
+// Home an axis
+void
+command_load_cell_endstop_home(uint32_t *args)
+{
+    struct load_cell_endstop *lce = load_cell_endstop_oid_lookup(args[0]);
+    sched_del_timer(&lce->time);
+    // clear the homing trigger flag
+    clear_flag(FLAG_IS_HOMING_TRIGGER, lce);
+    clear_flag(FLAG_IS_HOMING, lce);
+    lce->trigger_ticks = 0;
+    lce->ts = NULL;
+    // 0 samples indicates homing is finished
+    if (args[3] == 0) {
+        // Disable end stop checking
+        return;
+    }
+    lce->ts = trsync_oid_lookup(args[1]);
+    lce->trigger_reason = args[2];
+    lce->error_reason = args[3];
+    lce->time.waketime = args[4];
+    lce->rest_ticks = args[5];
+    lce->watchdog_max = args[6];
+    lce->watchdog_count = 0;
+    reset_filter_state(lce->sf);
+    lce->time.func = watchdog_event;
+    sched_add_timer(&lce->time);
+    set_flag(FLAG_IS_HOMING, lce);
```

Comment:
It's very strange to set the flag after enabling the timer.  It's enabling a race condition between task and interrupts.

Reply:
This flag is gone now